effis/src/ratelimit.rs

```diff
@@ -0,0 +1,206 @@
+use std::{
+    fmt::Display,
+    time::{Duration, SystemTime},
+};
+
+use rocket::{http::Header, response::Responder};
+use rocket_db_pools::{deadpool_redis::redis::AsyncCommands, Connection};
+use todel::{
+    models::{ErrorResponse, ErrorResponseData, FileSizeRatelimitedError, RatelimitError},
+    Conf,
+};
+
+use crate::Cache;
+
+pub type RatelimitedRouteResponse<T> =
+    Result<RatelimitHeaderWrapper<T>, RatelimitHeaderWrapper<ErrorResponse>>;
+
+/// The necessary headers for responses
+#[derive(Debug, Responder)]
+pub struct RatelimitHeaderWrapper<T> {
+    pub inner: T,
+    pub ratelimit_reset: Header<'static>,
+    pub ratelimit_max: Header<'static>,
+    pub ratelimit_bytes_limit: Header<'static>,
+    pub ratelimit_last_reset: Header<'static>,
+    pub ratelimit_request_count: Header<'static>,
+    pub ratelimit_sent_bytes: Header<'static>,
+}
+
+#[derive(Debug)]
+pub struct Ratelimiter {
+    key: String,
+    reset_after: Duration,
+    request_limit: u32,
+    file_size_limit: u64,
+    request_count: u32,
+    last_reset: u64,
+    sent_bytes: u64,
+}
+
+impl Ratelimiter {
+    /// Creates a new Ratelimiter
+    pub fn new<I>(bucket: &str, attachment_bucket: &str, identifier: I, conf: &Conf) -> Ratelimiter
+    where
+        I: Display,
+    {
+        let (reset_after, request_limit, file_size_limit) = match bucket {
```

Comment:
To reduce repetition maybe make this return a common struct with reset_after limit and file_size_limit?

Reply:
There is a struct for that, and this is constructing it